src/Common/PageCache.cpp

```diff
@@ -0,0 +1,665 @@
+                continue;
+            big_page_populated = true;
+            populated_pages += 1;
+
+            char & byte = chunk->data[idx * bytes_per_page];
+            chunk->first_bit_of_each_page.set(idx, (byte & 1) != 0);
+            byte |= 1;
+        }
+        if (big_page_populated)
+            populated_big_pages += 1;
+    }
+    int r = madvise(chunk->data, chunk->size, MADV_FREE);
+    if (r != 0)
+        logUnexpectedSyscallError("madvise(MADV_FREE)");
+
+    ProfileEvents::increment(ProfileEvents::PageCacheBytesUnpinnedRoundedToPages, bytes_per_page * populated_pages);
+    ProfileEvents::increment(ProfileEvents::PageCacheBytesUnpinnedRoundedToHugePages, bytes_per_page * pages_per_big_page * populated_big_pages);
+}
+
+std::pair<size_t, size_t> PageCache::restoreChunkFromLimbo(PageChunk * chunk, std::unique_lock<std::mutex> & /* chunk_mutex */) noexcept
+{
+    auto * data = const_cast<volatile char *>(chunk->data); // make sure our strategic memory reads/writes are not reordered or optimized out
+    size_t pages_restored = 0;
+    size_t pages_evicted = 0;
+    for (size_t idx = 0; idx < chunk->size / bytes_per_page; ++idx)
+    {
+        if (!chunk->pages_populated.get(idx))
+            continue;
+
+        /// After MADV_FREE, it's guaranteed that:
+        ///  * writing to the page makes it non-freeable again (reading doesn't),
+        ///  * after the write, the page contents are either fully intact or fully zero-filled,
+        ///  * even before the write, reads return either intact data (if the page wasn't freed) or zeroes (if it was, and the read page-faulted).
+        /// (And when doing the write there's no way to tell whether it page-faulted or not, AFAICT; that would make our life much easier!)
+        ///
+        /// With that in mind, we do the following dance to bring the page back from the MADV_FREE limbo:
+        ///  0. [in advance] Before doing MADV_FREE, make sure the page's first byte is not zero.
+        ///     We do it by setting the lowest bit of the first byte to 1, after saving the original value of that bit into a bitset.
+        ///  1. Read the second byte.
+        ///  2. Write the second byte back. This makes the page non-freeable.
+        ///  3. Read the first byte.
+        ///    3a. If it's zero, the page was freed.
+        ///        Set the second byte to 0, in case the page was freed between steps 1 and 2.
+        ///    3b. If it's nonzero, the page is intact.
+        ///        Restore the lowest bit of the first byte to the saved original value from the bitset.
+
+        char second_byte = data[idx * bytes_per_page + 1];
+        data[idx * bytes_per_page + 1] = second_byte;
+
+        char first_byte = data[idx * bytes_per_page];
+        if (first_byte == 0)
+        {
+            pages_evicted += 1;
+            data[idx * bytes_per_page + 1] = 0;
+            chunk->pages_populated.unset(idx);
+        }
```

Comment:
Ah, I meant "in case" as in "just in case" or "pack an umbrella in case it rains". We set the second byte to 0 unconditionally, which is useful if the page was freed between steps 1 and 2, and harmless otherwise, so we don't have to know (and indeed we can't know). This is just to keep unpopulated pages filled with zeroes. I'll update the comment.

Reply:
Ahh, makes sense! Thanks for the explanation!